File: python/docmost/http_client.py
"""HTTP transport with Bearer and cookie session auth."""
from __future__ import annotations

import threading
from typing import Any, Mapping, Optional
from urllib.parse import quote

import httpx

from docmost.exceptions import DocmostApiError
from docmost.models.api_response import ApiResponse, ErrorResponse
from docmost.models.login_dto import LoginDto


class DocmostHttpClient:
    def __init__(
        self,
        base_url: str,
        *,
        api_token: Optional[str] = None,
        email: Optional[str] = None,
        password: Optional[str] = None,
        timeout: float = 100.0,
        client: Optional[httpx.Client] = None,
    ) -> None:
        if api_token and (email or password):
            raise ValueError("Use either api_token or email/password, not both.")

        self._api_token = api_token
        self._email = email
        self._password = password
        self._logged_in = False
        self._login_lock = threading.Lock()

        api_base = base_url.rstrip("/") + "/api/"
        self._client = client or httpx.Client(
            base_url=api_base,
            timeout=timeout,
            follow_redirects=True,
        )
        self._owns_client = client is None

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> DocmostHttpClient:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    def ensure_authenticated(self) -> None:
        if self._api_token:
            return
        if not self._email or not self._password:
            raise RuntimeError("Provide api_token or email and password.")
        if self._logged_in:
            return
        with self._login_lock:
            if self._logged_in:
                return
            self._login()
            self._logged_in = True

    def login(self) -> None:
        """Authenticate with email/password (cookie session)."""
        if self._api_token:
            return
        with self._login_lock:
            self._login()
            self._logged_in = True

    def logout(self) -> ApiResponse[Any]:
        self.post("auth/logout", body=None)
        self._logged_in = False
        return ApiResponse[Any](data=None, success=True, status=200)

    def _login(self) -> None:
        payload = LoginDto(email=self._email or "", password=self._password or "")
        response = self._client.post(
            "auth/login",
            json=payload.model_dump(by_alias=True, exclude_none=True),
        )
        self._raise_for_status(response)
        envelope = ApiResponse[Any].model_validate(response.json())
        if not envelope.success:
            raise DocmostApiError("Login failed.", envelope.status)

    def _headers(self) -> dict[str, str]:
        if self._api_token:
            return {"Authorization": f"Bearer {self._api_token}"}
        return {}

    def post(self, path: str, *, body: Optional[Mapping[str, Any]]) -> ApiResponse[Any]:
        self.ensure_authenticated()
        response = self._client.post(
            path.lstrip("/"),
            json=body,
            headers=self._headers(),
        )
        return self._parse_envelope(response)

    def get(
        self,
        path: str,
        *,
        params: Optional[dict[str, str]] = None,
    ) -> ApiResponse[Any]:
        self.ensure_authenticated()
        response = self._client.get(
            path.lstrip("/"),
            params=params,
            headers=self._headers(),
        )
        return self._parse_envelope(response)

    def get_raw(
        self,
        path: str,
        *,
        params: Optional[dict[str, str]] = None,
    ) -> httpx.Response:
        self.ensure_authenticated()
        response = self._client.get(
            path.lstrip("/"),
            params=params,
            headers=self._headers(),
        )
        self._raise_for_status(response)
        return response

    def post_multipart(
        self,
        path: str,
        *,
        data: Optional[dict[str, str]],
        files: Optional[dict[str, tuple[str, Any, str]]],
    ) -> ApiResponse[Any]:
        self.ensure_authenticated()
        response = self._client.post(
            path.lstrip("/"),
            data=data,
            files=files,
            headers=self._headers(),
        )
        return self._parse_envelope(response)

    @staticmethod
    def _parse_envelope(response: httpx.Response) -> ApiResponse[Any]:
        DocmostHttpClient._raise_for_status(response)
        envelope = ApiResponse[Any].model_validate(response.json())
        if not envelope.success:
            raise DocmostApiError(
                f"API returned success=false (status {envelope.status}).",
                envelope.status,
            )
        return envelope

    @staticmethod
    def _raise_for_status(response: httpx.Response) -> None:
        if response.is_success:
            return
        error: Optional[ErrorResponse] = None
        try:
            error = ErrorResponse.model_validate(response.json())
        except Exception:
            pass
        message = (
            (error.error if error else None)
            or (str(error.message) if error and error.message else None)
            or response.reason_phrase
            or "Request failed."
        )
        raise DocmostApiError(message, response.status_code, error=error)

    @staticmethod
    def escape_path_segment(value: str) -> str:
        return quote(value, safe="")

File: typescript/src/http-client.ts
/** HTTP transport with Bearer and cookie session auth. */
import axios, { type AxiosInstance, type AxiosResponse, isAxiosError } from 'axios';
import { wrapper } from 'axios-cookiejar-support';
import { CookieJar } from 'tough-cookie';

import { DocmostApiError } from './errors.js';
import type { ApiResponse, ErrorResponse } from './types.js';
import type { LoginDto } from './models/login-dto.js';

export class DocmostHttpClient {
  private readonly apiToken?: string;
  private readonly email?: string;
  private readonly password?: string;
  private loggedIn = false;
  private readonly axios: AxiosInstance;

  constructor(
    baseUrl: string,
    options: {
      apiToken?: string;
      email?: string;
      password?: string;
      timeoutMs?: number;
    } = {},
  ) {
    if (options.apiToken && (options.email || options.password)) {
      throw new Error('Use either apiToken or email/password, not both.');
    }

    this.apiToken = options.apiToken;
    this.email = options.email;
    this.password = options.password;

    const jar = new CookieJar();
    const apiBase = baseUrl.replace(/\/$/, '') + '/api/';

    this.axios = wrapper(
      axios.create({
        baseURL: apiBase,
        timeout: options.timeoutMs ?? 100_000,
        jar,
        withCredentials: true,
        headers: { Accept: 'application/json' },
      } as Parameters<typeof axios.create>[0]),
    );
  }

  async ensureAuthenticated(): Promise<void> {
    if (this.apiToken) return;
    if (!this.email || !this.password) {
      throw new Error('Provide apiToken or email and password.');
    }
    if (this.loggedIn) return;
    await this.login();
    this.loggedIn = true;
  }

  async login(): Promise<void> {
    if (this.apiToken) return;
    const payload: LoginDto = { email: this.email!, password: this.password! };
    const response = await this.request(() =>
      this.axios.post<ApiResponse<unknown>>('auth/login', payload, {
        headers: { 'Content-Type': 'application/json' },
      }),
    );
    this.parseEnvelope(response);
    this.loggedIn = true;
  }

  async logout(): Promise<ApiResponse<unknown>> {
    const result = await this.post('auth/logout', undefined);
    this.loggedIn = false;
    return result;
  }

  async post<T = unknown>(path: string, body?: unknown): Promise<ApiResponse<T>> {
    await this.ensureAuthenticated();
    const response = await this.request(() =>
      this.axios.post<ApiResponse<T>>(path.replace(/^\//, ''), body, {
        headers: { ...this.authHeaders(), 'Content-Type': 'application/json' },
      }),
    );
    return this.parseEnvelope(response);
  }

  async get<T = unknown>(
    path: string,
    params?: Record<string, string>,
  ): Promise<ApiResponse<T>> {
    await this.ensureAuthenticated();
    const response = await this.request(() =>
      this.axios.get<ApiResponse<T>>(path.replace(/^\//, ''), {
        params,
        headers: this.authHeaders(),
      }),
    );
    return this.parseEnvelope(response);
  }

  async getRaw(path: string, params?: Record<string, string>): Promise<Response> {
    await this.ensureAuthenticated();
    const base = this.axios.defaults.baseURL ?? '';
    const url = new URL(path.replace(/^\//, ''), base);
    if (params) {
      for (const [k, v] of Object.entries(params)) {
        url.searchParams.set(k, v);
      }
    }
    const headers: Record<string, string> = { ...this.authHeaders() };
    const res = await fetch(url.toString(), { method: 'GET', headers });
    if (!res.ok) {
      await this.throwFromFetch(res);
    }
    return res;
  }

  async postMultipart<T = unknown>(
    path: string,
    request: Record<string, unknown>,
  ): Promise<ApiResponse<T>> {
    await this.ensureAuthenticated();
    const form = new FormData();
    for (const [key, value] of Object.entries(request)) {
      if (value === undefined || value === null) continue;
      if (
        typeof Blob !== 'undefined' &&
        (value instanceof Blob || (typeof Buffer !== 'undefined' && Buffer.isBuffer(value)))
      ) {
        const blob =
          value instanceof Blob ? value : new Blob([new Uint8Array(value as Buffer)]);
        form.append(key, blob, 'upload.bin');
      } else {
        form.append(key, String(value));
      }
    }
    const response = await this.request(() =>
      this.axios.post<ApiResponse<T>>(path.replace(/^\//, ''), form, {
        headers: { ...this.authHeaders(), 'Content-Type': 'multipart/form-data' },
      }),
    );
    return this.parseEnvelope(response);
  }

  private authHeaders(): Record<string, string> {
    if (this.apiToken) {
      return { Authorization: `Bearer ${this.apiToken}` };
    }
    return {};
  }

  private async request<T>(fn: () => Promise<AxiosResponse<T>>): Promise<AxiosResponse<T>> {
    try {
      return await fn();
    } catch (err) {
      if (isAxiosError(err) && err.response) {
        const data = err.response.data as ErrorResponse | undefined;
        const message =
          data?.error ??
          (typeof data?.message === 'string' ? data.message : undefined) ??
          err.message;
        throw new DocmostApiError(message, err.response.status, data);
      }
      throw err;
    }
  }

  private parseEnvelope<T>(response: AxiosResponse<ApiResponse<T>>): ApiResponse<T> {
    const envelope = response.data;
    if (!envelope) {
      throw new DocmostApiError('Empty response body.', response.status);
    }
    if (!envelope.success) {
      throw new DocmostApiError(
        `API returned success=false (status ${envelope.status}).`,
        envelope.status,
      );
    }
    return envelope;
  }

  private async throwFromFetch(res: Response): Promise<never> {
    let errorBody: ErrorResponse | undefined;
    try {
      errorBody = (await res.json()) as ErrorResponse;
    } catch {
      /* ignore */
    }
    const message =
      errorBody?.error ??
      (typeof errorBody?.message === 'string' ? errorBody.message : undefined) ??
      res.statusText ??
      'Request failed.';
    throw new DocmostApiError(message, res.status, errorBody);
  }
}
